core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -0,0 +1,214 @@
+         if (!info.isWriteOwner()) {
+            // topology changed, no longer an owner
+            remove(entry.getKey(), entry.getValue());
+            continue;
+         }
+         if (!info.isPrimary()) {
+            // backup owner
+            continue;
+         }
+         stage.dependsOn(new CleanupTask(this, entry.getKey(), entry.getValue()).checkRemoteSites());
+      }
+      return stage.freeze();
+   }
+
+   private void remove(Object key, TombstoneData data) {
+      tombstoneMap.remove(key, data);
+   }
+
+   private Collection<XSiteBackup> getAsyncBackups() {
+      return asyncBackups;
+   }
+
+   private RpcManager getRpcManager() {
+      return rpcManager;
+   }
+
+   private CommandsFactory getCommandsFactory() {
+      return commandsFactory;
+   }
+
+   private DistributionInfo getSegmentDistribution(int segment) {
+      return distributionManager.getCacheTopology().getSegmentDistribution(segment);
+   }
+
+   private static class TombstoneData {
+      private final int segment;
+      private final IracMetadata metadata;
+
+      private TombstoneData(int segment, IracMetadata metadata) {
+         this.segment = segment;
+         this.metadata = Objects.requireNonNull(metadata);
+      }
+
+      public int getSegment() {
+         return segment;
+      }
+
+      public IracMetadata getMetadata() {
+         return metadata;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+         if (this == o) return true;
+         if (o == null || getClass() != o.getClass()) return false;
+         TombstoneData that = (TombstoneData) o;
+         return metadata.equals(that.metadata);
+      }
+
+      @Override
+      public int hashCode() {
+         return metadata.hashCode();
+      }
+   }
+
+   private static class CleanupTask implements Function<Boolean, CompletionStage<Void>>, Runnable {
+      private final DefaultIracTombstoneManager cleaner;
+      private final Object key;
+      private final TombstoneData tombstone;
+
+      private CleanupTask(DefaultIracTombstoneManager cleaner, Object key, TombstoneData tombstone) {
+         this.cleaner = cleaner;
+         this.key = key;
+         this.tombstone = tombstone;
+      }
+
+      CompletionStage<Void> checkRemoteSites() {
+         AggregateCompletionStage<Boolean> stage = CompletionStages.orBooleanAggregateCompletionStage();
+         for (XSiteBackup backup : cleaner.getAsyncBackups()) {
+            // we don't need the tombstone to query the remote site
+            IracCleanupTombstoneCommand cmd = cleaner.getCommandsFactory().buildIracCleanupTombstoneCommand(key, null);
+            stage.dependsOn(cleaner.getRpcManager().invokeXSite(backup, cmd)
+                  .exceptionally(CompletableFutures.toTrueFunction())); // in case of exception, keep the tombstone
+         }
+         return stage.freeze().thenCompose(this);
```

Comment:
```suggestion
            stage.dependsOn(cleaner.getRpcManager().invokeXSite(backup, cmd);
         }
         // in case of exception, keep the tombstone
         return stage.freeze()
                     .exceptionally(CompletableFutures.toTrueFunction()))
                     .thenCompose(this);
```

And shouldn't we log the failures?

Reply:
no, they can become quite verbose and not useful IMO.